ModernGov template page generation module

Adnan-cds opened the pull request:
Bringing in the ModernGov template page module from the Croydon site.

ekes commented:
Just trying to learn the architecture here. The first question that springs to mind here is: If I remember this "moderngov" thing works with a single URL? What's the reason to use a node which seems to add lots of complexity, rather than just including a controller?

Adnan-cds commented:
> What's the reason to use a node which seems to add lots of complexity, rather than just including a controller?

Hi Ekes, thanks for being the first person to look into this :)  Yes, a controller can do the job too.  I thought the level of complexity is not very different, so just used a node.  I also thought nodes have two minor advantages:
- Easier for editors to publish/unpublish and update the path alias during development stage.
- We can use all the node related APIs which could be useful to people.  Particularly given that we don't know who is going to use this module and what their requirement is.

But these are very insignificant benefits.  If you prefer a controller, I can swap out the node with a controller.  Just let me know.

The biggest effort has gone into figuring out how to render all the URLs as absolute URLs.  I wasn't feeling comfortable doing a regex replacement of URLs, so had to do it using a custom Path processor service.  I also had to [convince](https://github.com/localgovdrupal/localgov_moderngov_tpl/blob/369273b86e438efd3ebbed075ef7dc598c85ced2/src/EventSubscriber/KernelRequestSubscriber.php#L58) ```file_url_transform_relative()``` to stand by.  That's all this module does.

ekes commented:
>  I thought the level of complexity is not very different, so just used a node

Ok.

> Easier for editors to publish/unpublish and update the path alias during development stage.

Actually that's a point. Path is actually something you want to deploy isn't it. Like it wants to be fixed by a developer who also configures it in ModernGov, and if it then changes later it breaks?

> The biggest effort has gone into figuring out how to render all the URLs as absolute URLs.

I'll have a look. Guess it just needs some solid tests. For all the places, and ways, a URL could appear. Understand the reluctance to just regex the output, but it should cover all the unforeseen cases.

finnlewis commented:
@Adnan-cds - we were just taking a look at this at 'Merge Monday'. 

Looks like @ekes is already reviewing with you.

@andybroomfield is also interested and will take a look if he can.

Adnan-cds commented:
> For all the places, and ways, a URL could appear. Understand the reluctance to just regex the output, but it should cover all the unforeseen cases.

That's right.  So the code in this pull request only covers the menu blocks and Alert banner block.  This is as long as Alert banners are using LinkIt links rather than hardcoded relative links.  This works for the Croydon site but can't be enough for everyone.  If this change is accepted, I will have to raise a follow up pull request to make the *list* of such blocks configurable through a config form.

The biggest trouble so far has been with the hardcoded URLs in Twig templates.  I had to raise a [pull request for the localgov_theme](https://github.com/localgovdrupal/localgov_theme/pull/197) to resolve this and Steve pointed out that we are moving away from the localgov_theme to the localgov_base theme.  So themes can continue to cause issues.  I am thinking of adding some sample code in the README to help avoid this.

> we were just taking a look at this at 'Merge Monday'.

Sorry Finn, I missed it again :(

finnlewis commented:
How's this one coming along @Adnan-cds @ekes ?

andybroomfield commented:
After following the steps to test, I get this:
<img width="1401" alt="Screenshot 2022-01-17 at 2 39 39 pm" src="https://user-images.githubusercontent.com/1467480/149789782-276ff736-36ed-4b29-b840-543b52f9e678.png">

And as an anon user
<img width="1401" alt="Screenshot 2022-01-17 at 2 39 36 pm" src="https://user-images.githubusercontent.com/1467480/149789844-01615c03-78f4-4e76-a1e7-df4b03a4773a.png">

This is on fresh install with localgov demo content enabled.

How it looks without normally.
<img width="1401" alt="Screenshot 2022-01-17 at 2 46 27 pm" src="https://user-images.githubusercontent.com/1467480/149790701-2eec98ab-7200-470b-9bc4-ef130b77e308.png">

Links do render as absolute though.

Adnan-cds commented:
Hi Andy,
So this is to do with the page template for the ModernGov page.  The localgov_moderngov_tpl module provides a [generic one](https://github.com/localgovdrupal/localgov_moderngov_tpl/blob/feature/first-cut/templates/page--moderngov-page-tpl.html.twig) that doesn't go well with the localgov_base/localgov_scarfolk theme at all.  I have modified the generic page template markup slightly and dropped inside **localgov_scarfolk** to get a [passable output ](https://user-images.githubusercontent.com/50206849/149825064-aa78bac0-7634-44df-a9c4-ab5b5e4e1165.png) .  This is what I have added to **localgov_scarfolk/templates/layout/page--moderngov-page-tpl.html.twig**:

```
<div class="layout-container">

  <header class="lgd-header" role="banner">
    <div class="lgd-container">
      {{ page.header }}
    </div>
  </header>

  {{ page.primary_menu }}
  {{ page.secondary_menu }}

  <div class="breadcrumb-wrapper lgd-container">
    {breadcrumb}
  </div>

  {{ page.highlighted }}

  {{ page.help }}

  {# Site admins should be able to edit/delete the ModernGov template page. #}
  {% if is_admin and page.content_top %}
    {{ page.content_top }}
  {% endif %}

  <main class="lgd-container" role="main">
    <a id="main-content" tabindex="-1"></a>

    <div class="lgd-row">
      <div class="layout-content lgd-row__two-thirds">
        {# Site admins should be able to edit/delete the ModernGov template page. #}
        {% if is_admin %}
          {{ page.content }}
        {% else %}
          {content}
        {% endif %}
      </div>

      {% if page.sidebar_second %}
        <aside class="layout-sidebar-second lgd-row__one-thirds" role="complementary">
          {sidenav}
        </aside>
      {% endif %}
    </div>

  </main>

  {% if page.footer_second %}
    <footer class="lgd-footer lgd-footer__pre-footer" role="contentinfo">
      {{ page.footer_second }}
    </footer>
  {% endif %}

</div>
```

I can see that the [README](https://github.com/localgovdrupal/localgov_moderngov_tpl/blob/feature/first-cut/README.md) merits a section on page template customization.

finnlewis commented:
@andybroomfield has yet to test this again.

andybroomfield commented:
Having followed @Adnan-cds instructions I see the following.
<img width="1401" alt="Screenshot 2022-01-24 at 2 53 22 pm" src="https://user-images.githubusercontent.com/1467480/150805996-2b5c6813-48d3-4ba4-9460-079aead0a0b0.png">

finnlewis commented:
Just a note here to say that we might want to provide the template as part of the localgov_base theme. 

https://github.com/localgovdrupal/localgov_base/issues/174

andybroomfield commented:
Logged out screen.
<img width="1401" alt="Screenshot 2022-01-24 at 2 55 48 pm" src="https://user-images.githubusercontent.com/1467480/150806442-736f3077-7cf7-4af8-a440-aeda7b96b598.png">

Adnan-cds commented:
This module now only works with Drupal 9.3.x which has [replaced](https://www.drupal.org/node/2940031) file_uri_transform_relative() with a new service called file_url_generator.  This change makes the code of this module simpler but at the cost of loosing compatibility with Drupal 9.2.x and below.  I hope that's okay.

Also noticed that the localgov profile is failing [deprecation code checks](https://github.com/localgovdrupal/localgov_moderngov_tpl/runs/4955743455?check_suite_focus=true#step:4:20) and as such all test runs are now failing :( @finnlewis

stephen-cox commented:
> Also noticed that the localgov profile is failing deprecation code checks and as such all test runs are now failing :( 

PR for failing tests here: https://github.com/localgovdrupal/localgov/pull/420

markconroy commented:
[mdgov.zip](https://github.com/localgovdrupal/localgov_moderngov_tpl/files/8016342/mdgov.zip)

Any time I have implemented MD Gov, I've used a controller and set the specific URL with the routing.yml file.

I don't have any issue with the node approach (actually, it's very clever), but I like the controller approach for a distribution for 2 reasons:

1. We force the same URL for each council and this helps us with consistency - I'm not sure if the node approach will break in the future, but anything that introduces inconsistencies when you are trying to do things at scale has the potential to come back and harm us
2. It removes the ability for editors to change/add/edit content on the page. This page has one job to do (and serving custom content from the CMS is not that job). When we render a specific template, we can then remove all the regions except the header, main, sidebar, breadcrumbs, and footer. This significantly reduced the amount of non-absolute URLs we might need to deal with.

I've attached a tiny, simple module that will position our mdgov templates at example.com/mdgov and all we need to do then is add the page.html.twig template to our localgov_base theme (and duplicate it in this module if we wish).

markconroy commented:
You know what? Having read through all the code in the module, and how we have a rewriting system for the URLs (though I think we might need to finish that to work with the CSS/JS assets), I think @Adnan-cds's work on this is very strong, and should probably be the approach we take.

There's some very minor things we could update, such as removing classes with underscore and using hyphens instead, and changing the template name to `page--moderngov.html.twig`, but these are very minor issues.

markconroy commented on composer.json:
```diff
@@ -0,0 +1,10 @@
+{
+    "name": "localgovdrupal/localgov_moderngov_tpl",
```
I think we should call the module `localgov_moderngov` or `localgov_moderngov_template` and not have abbreviations such as `tpl` which is more of a D7 thing.

Adnan-cds replied:
Hmm.. not a bad suggestion but that entails lots of changes.  Any preference Finn? @finnlewis

markconroy commented on localgov_moderngov_tpl.module:
```diff
@@ -0,0 +1,93 @@
+
+  $is_moderngov_tpl_page = Drupal::service('localgov_moderngov_tpl.moderngov_path_predicate')->isModernGovTplPage();
+  if ($is_moderngov_tpl_page) {
+    $vars['attributes']['class'][] = 'node--localgov_modern_gov_page';
```
Drupal coding standards requires hyphens/dashes instead of underscores to separate words in class names. Underscores are only for specifying elements level items inside (BEM) blocks - such as `node` with a `node__content` or `node__title` element inside it.

Since this class is on the body element, it should be page level, not node level. I suggest we rename it to `page--moderngov` or even simply `page-modern-gov`. It might also be good to have a hook_preprocess_node to add `node--moderngov` to the node for this page.

Adnan-cds replied:
I have renamed the ```node--localgov_modern_gov_page``` classname to ```page--moderngov-template```.  Will that do?

markconroy commented on localgov_moderngov_tpl.module:
```diff
@@ -0,0 +1,93 @@
+ */
+function localgov_moderngov_tpl_block_build_alter(array &$build, BlockPluginInterface $block_plugin) {
+  $block_type = $block_plugin->getBaseId();
+  $is_moderngov_page_block = in_array($block_type, [
```
Is there anyway we can add "all" blocks to this array, so any site that has a custom block does not break?

Adnan-cds replied:
That's actually a [todo](https://github.com/localgovdrupal/localgov_moderngov_tpl/blob/feature/first-cut/README.md#todo) that I hope to implement in a different ticket.  So the idea is to add a config form where you manually select the blocks that appear in the ModernGov template page.  That way, we only add the cache context data to the relevant blocks and leave the others undisturbed.  Saves some storage space :)

If I knew what blocks are to appear on the ModernGov template page then we wouldn't even need a config form.  I will surely explore that approach when it's time to implement this.

markconroy commented on src/FileUrlGeneratorDecorator.php:
```diff
@@ -0,0 +1,90 @@
+  /**
+   * {@inheritdoc}
+   *
+   * When on the ModernGov tpl page, generates absolute URLs even for relative
```
I LOVE this. Thanks!

Adnan-cds replied:
Yeah, this file_url_generator service is a [new feature](https://www.drupal.org/node/2940031) of Drupal 9.3.x that has made things so much easier for this ModernGov thing :)

markconroy commented on templates/page--moderngov-page-tpl.html.twig:
```diff
@@ -0,0 +1,99 @@
+      {% endif %}
+    </div>
+
+    {% if page.sidebar_second %}
```
I think we should remove the `if` around this and force the sidebar. If not, the moderngov sidebar will not be visible unless someone has placed a block in there. We can't be certain of this.

Adnan-cds replied:
Done.  I should have thought of this.

Adnan-cds commented:
> ... and changing the template name to page--moderngov.html.twig, but these are very minor issues.

I have renamed the page template to page--moderngov-template.html.twig.  Let me know if that's acceptable. @markconroy

Adnan-cds commented:
I stumbled across this the other day and this appears to be a simpler way to generate absolute URLs - [Drupal\Component\Utility\Html\transformRootRelativeUrlsToAbsolute()](https://git.drupalcode.org/project/drupal/-/blob/9.4.x/core/lib/Drupal/Component/Utility/Html.php#L454) - I am thinking of marrying this up with Eke's original ModernGov code in favour of what I have written so far.
